packages/fault-detector/src/service.ts

```diff
@@ -0,0 +1,179 @@
+          type: Gauge,
+          desc: 'Highest good batch index',
+        },
+        highestKnownBatchIndex: {
+          type: Gauge,
+          desc: 'Highest known batch index',
+        },
+        isCurrentlyMismatched: {
+          type: Gauge,
+          desc: '0 if state is ok, 1 if state is mismatched',
+        },
+        inUnexpectedErrorState: {
+          type: Gauge,
+          desc: '0 if service is ok, 1 service is in unexpected error state',
+        },
+      },
+    })
+  }
+
+  async init(): Promise<void> {
+    const network = await this.options.l1RpcProvider.getNetwork()
+    this.state.messenger = new CrossChainMessenger({
+      l1SignerOrProvider: this.options.l1RpcProvider,
+      l2SignerOrProvider: this.options.l2RpcProvider,
+      l1ChainId: network.chainId,
+    })
+
+    this.state.highestCheckedBatchIndex = this.options.startBatchIndex
+  }
+
+  async main(): Promise<void> {
+    const latestBatchIndex =
+      await this.state.messenger.contracts.l1.StateCommitmentChain.getTotalBatches()
+    if (this.state.highestCheckedBatchIndex >= latestBatchIndex.toNumber()) {
+      await sleep(15000)
+      return
+    }
+
+    this.metrics.highestKnownBatchIndex.set(latestBatchIndex.toNumber())
+
+    this.logger.info(`checking batch`, {
+      batchIndex: this.state.highestCheckedBatchIndex,
+    })
+
+    const targetEvents =
+      await this.state.messenger.contracts.l1.StateCommitmentChain.queryFilter(
+        this.state.messenger.contracts.l1.StateCommitmentChain.filters.StateBatchAppended(
+          this.state.highestCheckedBatchIndex
+        )
+      )
+
+    if (targetEvents.length === 0) {
+      this.logger.error(`unable to find event for batch`, {
+        batchIndex: this.state.highestCheckedBatchIndex,
+      })
+      this.metrics.inUnexpectedErrorState.set(1)
+      return
+    }
+
+    if (targetEvents.length > 1) {
+      this.logger.error(`found too many events for batch`, {
+        batchIndex: this.state.highestCheckedBatchIndex,
+      })
+      this.metrics.inUnexpectedErrorState.set(1)
+      return
+    }
+
+    const targetEvent = targetEvents[0]
+    const batchTransaction = await targetEvent.getTransaction()
+    const [stateRoots] =
+      this.state.messenger.contracts.l1.StateCommitmentChain.interface.decodeFunctionData(
+        'appendStateBatch',
+        batchTransaction.data
+      )
+
+    const batchStart = targetEvent.args._prevTotalElements.toNumber() + 1
+    const batchSize = targetEvent.args._batchSize.toNumber()
+
+    // `getBlockRange` has a limit of 1000 blocks, so we have to break this request out into
+    // multiple requests of maximum 1000 blocks in the case that batchSize > 1000.
+    let blocks: any[] = []
+    for (let i = 0; i < batchSize; i += 1000) {
+      const provider = this.options
+        .l2RpcProvider as ethers.providers.JsonRpcProvider
+      blocks = blocks.concat(
+        await provider.send('eth_getBlockRange', [
```

Comment:
This will break when we release bedrock since this RPC is being removed, so if you are ok with that then we can leave this but it would be more ideal to use a batch json rpc request, ethers does support this

Reply:
Fixed

Reply:
Let's use this version for now, I will update when necessary